### DISCUSSION TITLE
timeline_variables as dynamic function

### DISCUSSION THREAD

**Discussion by patiencerosestevens** [jspsych/jsPsych#2598] [Q&A] [answered]
Hello! I have a question: is it possible to set `timeline_variables` as a dynamic parameter? I would like to be able to use a function to select the proper stimulus list, which is selected based on a participant's ID which they enter on a previous `survey-text` trial. I wrote the following code, using a variable `cond` which was defined as a global variable and then assigned a new value in a `on_finish` function during a previous trial: 

```
var bk3 = {
        timeline_variables: function () { return ld_stim_ops[cond] }, // also doesn't work with {return ld_stim_ops[2]}, does work with ld_stim_ops[1]
        randomize_order: true,
        timeline: [
            { // this event shows the sentence until the participant presses spacebar
                type: 'html-keyboard-response',
                stimulus: function () {
                    var stim = '<p style="font-size:22px;">' + jsPsych.timelineVariable('word') + ' ' + jsPsych.timelineVariable('stem') + '</p>';
                    return stim;
                },
                choices: ['f', 'j'],
                trial_duration: response_dur
            },
            { // this event shows the sentence until the participant presses spacebar
                type: 'html-keyboard-response',
                stimulus: ' ',
                trial_duration: iti,
                choices: ["NO_KEYS"]
            },
        ],
        on_start: function () { $('body').css('cursor', 'none') }, // this code makes the cursor (mouse) invisible during trials.
        on_finish: function () { $('body').css('cursor', 'default') }
    };
    timeline.push(bk3)
```

The error I get whenever I use any function to generate a value for `timeline_variables`, even if it doesn't involve the global variable `cond`, is `Uncaught TypeError TypeError: Cannot read properties of undefined (reading 'word')`. What can I do to fix this? I've also attached a script where this error is replicated below. Thank you very much for your help!
[probdemohtml.txt](https://github.com/jspsych/jsPsych/files/8525967/probdemohtml.txt)

**Comment by jodeleeuw** [comment, marked as answer]
Hi @patiencerosestevens,

> is it possible to set timeline_variables as a dynamic parameter?

Unfortunately it is not possible right now, though it is on our todo list for version 8. 

A potential workaround for now is to use the `conditional_function` on a timeline to only run the timeline with the appropriate version of the `timeline_variables`. Obviously this is not an ideal solution, and is only feasible if the number of conditions is relatively low or if you build the timelines using a loop.

**Reply by patiencerosestevens** [reply]
I see, thank you! I will use the conditional_function set up :-)
